include/cuda/memory_resource

```diff
@@ -0,0 +1,529 @@
+
+/// \concept async_resource
+/// \brief The \c async_resource concept
+template <class _Resource>
+_LIBCUDACXX_CONCEPT async_resource =
+  resource<_Resource>&& _Has_member_allocate_async<_Resource>&& _Has_member_deallocate_async<_Resource>;
+
+/// \concept has_property
+/// \brief The \c has_property concept
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_property_,
+                             requires(const _Resource& __res) //
+                             (get_property(__res, _Property{})));
+template <class _Resource, class _Property>
+_LIBCUDACXX_CONCEPT has_property = _LIBCUDACXX_FRAGMENT(_Has_property_, _Resource, _Property);
+
+template <class T>
+using _Has_value_types = typename T::value_type;
+
+/// \concept property_with_value
+/// \brief The \c property_with_value concept
+template <class _Property>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Property_with_value_,
+                             requires() //
+                             (typename(_Has_value_types<_Property>)));
+template <class _Property>
+_LIBCUDACXX_CONCEPT property_with_value = _LIBCUDACXX_FRAGMENT(_Property_with_value_, _Property);
+
+/// \concept has_property_with
+/// \brief The \c has_property_with concept
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT_FRAGMENT(_Has_property_with_,
+                             requires(const _Resource& __res) //
+                             (_CUDA_VSTD::_Returns_exactly<_Return>(get_property(__res, _Property{}))));
+template <class _Resource, class _Property, class _Return>
+_LIBCUDACXX_CONCEPT has_property_with = property_with_value<_Property>&&
+  _LIBCUDACXX_FRAGMENT(_Has_property_with_, _Resource, _Property, _Return);
+
+/// \concept resource_with
+/// \brief The \c resource_with concept
+template <class _Resource, class... _Properties>
+#if _LIBCUDACXX_STD_VER < 17
+_LIBCUDACXX_CONCEPT resource_with =
+  resource<_Resource>&& _CUDA_VSTD::conjunction_v<_CUDA_VSTD::bool_constant<has_property<_Resource, _Properties>>...>;
+#else
+_LIBCUDACXX_CONCEPT resource_with       = resource<_Resource> && (has_property<_Resource, _Properties> && ...);
+#endif
+
+/// \concept async_resource_with
+/// \brief The \c async_resource_with concept
+template <class _Resource, class... _Properties>
+#if _LIBCUDACXX_STD_VER < 17
+_LIBCUDACXX_CONCEPT async_resource_with = async_resource<_Resource>&&
+  _CUDA_VSTD::conjunction_v<_CUDA_VSTD::bool_constant<has_property<_Resource, _Properties>>...>;
+#else
+_LIBCUDACXX_CONCEPT async_resource_with = async_resource<_Resource> //
+                                          && (has_property<_Resource, _Properties> && ...);
+#endif
+
+///////////////////////////////////////////////////////////////////////////////
+/// class resource_ref
+/// class async_resource_ref
+enum class _AllocType
+{
+  _Default,
+  _Async,
+};
+
+struct _Alloc_vtable
```

Comment:
Currently we never use this alone. 

Do we want to change the code that we conditionally only use the smaller vtable for `resource_ref` or go the simple route and only ever use `_Async_alloc_vtable ` and remove the base class?

Reply:
I'd go for dropping conditionality entirely and template basic_resource_ref against alloc_base, not an enum. This keeps the system open, while not adding any complexity to it.